packaging/cmake/Modules/NetdataJSONC.cmake

```diff
@@ -0,0 +1,102 @@
+        # values instead of treating them as booleans, so we need to use
+        # proper strings for option values instead of just setting them
+        # to true or false.
+        set(DISABLE_BSYMBOLIC ON)
+        set(DISABLE_WERROR ON)
+        set(DISABLE_EXTRA_LIBS ON)
+        set(BUILD_SHARED_LIBS OFF)
+        set(BUILD_STATIC_LIBS ON)
+        set(BUILD_APPS OFF)
+
+        FetchContent_Declare(json-c
+                GIT_REPOSITORY https://github.com/json-c/json-c
+                GIT_TAG b4c371fa0cbc4dcbaccc359ce9e957a22988fb34 # json-c-0.17-20230812
+        )
+
+        FetchContent_MakeAvailable_NoInstall(json-c)
+
+        message(STATUS "Finished preparing vendored copy of JSON-C")
+endfunction()
+
+# Handle setup of json-c for the build.
+#
+# This will attempt to find json-c using pkg_check_modules. If it finds
+# a usable copy, that will be used. If not, it will bundle a vendored copy
+# as a sub-project.
+#
+# Irrespective of how json-c is to be included, library names,
+# include directories, and compile definitions will be specified in the
+# NETDATA_JSONC_* variables for later use.
+macro(netdata_detect_jsonc)
+        if(NOT ENABLE_BUNDLED_JSONC)
+                pkg_check_modules(JSONC json-c)
+        endif()
+
+        if(NOT JSONC_FOUND)
+                netdata_bundle_jsonc()
+                set(NETDATA_JSONC_LDFLAGS json-c)
+                set(NETDATA_JSONC_INCLUDE_DIRS ${PROJECT_BINARY_DIR}/include)
+                get_target_property(NETDATA_JSONC_CFLAGS_OTHER json-c INTERFACE_COMPILE_DEFINITIONS)
+
+                if(NETDATA_JSONC_CFLAGS_OTHER STREQUAL NETDATA_JSONC_CFLAGS_OTHER-NOTFOUND)
+                        set(NETDATA_JSONC_CFLAGS_OTHER "")
+                endif()
+
+                add_custom_command(
+                        OUTPUT ${PROJECT_BINARY_DIR}/include/json-c
+                        COMMAND ${CMAKE_COMMAND} -E make_directory ${PROJECT_BINARY_DIR}/include
+                        COMMAND ${CMAKE_COMMAND} -E create_symlink ${json-c_BINARY_DIR} ${PROJECT_BINARY_DIR}/include/json-c
+                        COMMENT "Create compatibility symlink for vendored JSON-C headers"
+                        DEPENDS json-c
+                )
+                add_custom_target(
+                        json-c-compat-link
+                        DEPENDS ${PROJECT_BINARY_DIR}/include/json-c
+                )
+        else()
+                set(NETDATA_JSONC_LDFLAGS ${JSONC_LDFLAGS})
+                set(NETDATA_JSONC_CFLAGS_OTHER ${JSONC_CFLAGS_OTHER})
+                set(NETDATA_JSONC_INCLUDE_DIRS ${JSONC_INCLUDE_DIRS})
+                add_custom_target(json-c-compat-link)
+        endif()
+endmacro()
+
+# Add json-c as a public link dependency of the specified target.
+#
+# The specified target must already exist, and the netdata_detect_json-c
+# macro must have already been run at least once for this to work correctly.
+function(netdata_add_jsonc_to_target _target)
```

Comment:
I’m not sure how this significantly complicates things, unless you’re using some external script to parse the CMakeLists file and try to figure out the dependency graph from that.

This function being called on a target says quite clearly that that target depends on JSON-C. And because of how CMake handles adding dependencies, you need to be searching the whole file for the target name anyway instead of just blindly looking for `target_*` calls on the target. I’m also trying to follow a `netdata_add_*_to_target` pattern for functions like this so that they are easy to find and properly descriptive.

To clarify, my intent here is to keep all the arcana involved in getting this to actually work properly encapsulated. By handling it like this, anybody working on the build system won’t need to remember the names of the variables used to add JSON-C to a target, or where they are set, or whether the include directories need to be appended or prepended to the search list, or the fact that you need to add a supplementary dependency to the target to ensure that headers can be found correctly when using a vendored copy of JSON-C.